AIM: To determine error/loss for feed forward network with 3 layers.

Bhavya Singhal

B1

B071

BTech CE

In [ ]:
#importing required libraries
import tensorflow as tf
import numpy as np
import pandas as pd


In [ ]:
import matplotlib.pyplot as plt

In [ ]:
df= pd.read_csv('/content/iris_data.csv')

In [ ]:
df.shape

(150, 5)

In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [ ]:
#allocating number to species
df1 = pd.get_dummies(df['species'])
df1.head()

Iris-setosa  Iris-versicolor  Iris-virginica
0            1                0               0
1            1                0               0
2            1                0               0
3            1                0               0
4            1                0               0

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
std_scaler = StandardScaler()

In [ ]:
# x= (x-mean)/(std deviation) for all feature values
df.iloc[:, [0,1,2,3]] = std_scaler.fit_transform(df.iloc[:,[0,1,2,3]])

In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width      species
0     -0.900681     1.032057     -1.341272    -1.312977  Iris-setosa
1     -1.143017    -0.124958     -1.341272    -1.312977  Iris-setosa
2     -1.385353     0.337848     -1.398138    -1.312977  Iris-setosa
3     -1.506521     0.106445     -1.284407    -1.312977  Iris-setosa
4     -1.021849     1.263460     -1.341272    -1.312977  Iris-setosa

In [ ]:
df.drop('species', axis=1, inplace=True)

In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width
0     -0.900681     1.032057     -1.341272    -1.312977
1     -1.143017    -0.124958     -1.341272    -1.312977
2     -1.385353     0.337848     -1.398138    -1.312977
3     -1.506521     0.106445     -1.284407    -1.312977
4     -1.021849     1.263460     -1.341272    -1.312977

In [ ]:
final = pd.concat([df,df1], axis=1)

In [ ]:
final.head()

sepal_length  sepal_width  petal_length  petal_width  Iris-setosa  \
0     -0.900681     1.032057     -1.341272    -1.312977            1   
1     -1.143017    -0.124958     -1.341272    -1.312977            1   
2     -1.385353     0.337848     -1.398138    -1.312977            1   
3     -1.506521     0.106445     -1.284407    -1.312977            1   
4     -1.021849     1.263460     -1.341272    -1.312977            1   

   Iris-versicolor  Iris-virginica  
0                0               0  
1                0               0  
2                0               0  
3                0               0  
4                0               0

In [ ]:
np.random.seed(42)
#4 rows 1 column
#from input layer (4 input nodes) to hidden layer (2 hidden layer nodes)
w_i_h1 = np.random.rand(4,1)
w_i_h2 = np.random.rand(4,1)
w_i_h3 = np.random.rand(4,1)

#3 output nodes
w_h_o1 = np.random.rand(3,1)
w_h_o2 = np.random.rand(3,1)
w_h_o3 = np.random.rand(3,1)

#random bias
b1=np.random.rand(1)
b2=np.random.rand(1)
b3=np.random.rand(1)

In [ ]:
w_i_h1

array([[0.37454012],
       [0.95071431],
       [0.73199394],
       [0.59865848]])

In [ ]:
def sig(x):
  sig_x = 1/(1+np.exp(-x))
  return sig_x

In [ ]:
input = final.iloc[:,[0,1,2,3]]

#2nd layer 1st node
z2_1 = np.dot(input, w_i_h1) + b1
#2nd layer 2nd node
z2_2 = np.dot(input, w_i_h2) + b2
#2nd layer 3rd node
z2_3 = np.dot(input, w_i_h3) + b3

In [ ]:
h2_1 = sig(z2_1)
h2_2 = sig(z2_2)
h2_3 = sig(z2_3)

In [ ]:
#2nd layer both nodes
h2 = np.append(h2_1, h2_2, axis=1)
h2 = np.append(h2, h2_3, axis=1)

In [ ]:
h2[0]

array([0.27200302, 0.28852507, 0.32183357])

In [ ]:
#output layer or 3rd layer
o3_1 = np.dot(h2, w_h_o1)
o3_2 = np.dot(h2, w_h_o2)
o3_3 = np.dot(h2, w_h_o3)

In [ ]:
#getting output
o1 = sig(o3_1)
o2 = sig(o3_2)
o3 = sig(o3_3)

In [ ]:
print(o1[0])
print(o2[0])
print(o3[0])

[0.58569808]
[0.5760435]
[0.59831089]


In [ ]:
final.iloc[0, :]

sepal_length      -0.900681
sepal_width        1.032057
petal_length      -1.341272
petal_width       -1.312977
Iris-setosa        1.000000
Iris-versicolor    0.000000
Iris-virginica     0.000000
Name: 0, dtype: float64

In [ ]:
#creating numpy array of all possible outputs
f1 = np.array(final.iloc[:,4])
f2 = np.array(final.iloc[:,5])
f3 = np.array(final.iloc[:,6])

#reshaping array such that all elements are put in one column
f1=f1.reshape((-1,1))
f2=f2.reshape((-1,1))
f3=f3.reshape((-1,1))

In [ ]:
#calculating error for all predicted outputs
e1 = f1-o1
e2 = f2-o2
e3 = f3-o3

In [ ]:
#calculating sum of squared errors
error = np.multiply(e1,e1) + np.multiply(e2,e2) + np.multiply(e3,e3)

In [ ]:
total_error = np.sum(error)

In [ ]:
print(total_error)

147.44942573200146


Conclusion:

For the given iris dataset, neural network with 3 layers is built. Hidden layer has 2 neurons and sigmoid is used as activation function. For above architecture with forward propagation the error/loss is 144.83, when number of neurons is increased in hidden layer from 2 to 3 the error/loss is increased to 147.44942573200146. The above result shows that error does not always reduce with increase in number of neurons. Error can only be reduced by using back propagation.